### PR TITLE
LocalModelStorage bypasses windows path name length restriction upon saving archive

### PR DESCRIPTION
**Pull request by m-vdb**
**Proposed changes**:
- LocalModelStorage bypasses windows path name length restriction upon saving archive
- Made sure the test was failing using TDD (see [build logs](https://github.com/RasaHQ/rasa/actions/runs/3564189533/jobs/5987876826))

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by m-vdb** [comment]
@mvielkind would you care to share how you tested the long file path fix using `\\\\?\\`? I tried different things, and it seems that the built-in `open()` method doesn't even recognise the prefix when trying to open the file and write into it, and I get an `OSError: [Errno 22] Invalid argument`. You can see the error in the CI job with the logs there. I was able to reproduce the same error from a Windows VM, but couldn't find a way to fix it; and of course I couldn't find documentation of the issue out there....

**Comment by m-vdb** [comment]
actually I hit another limit, described [here](https://github.com/RasaHQ/rasa/pull/11792/files#diff-d19e7f0f0a859c44b0a1454f7b69cabd99e0aa256a3a45b651516aee98f220acR100-R102). Each path component on windows cannot have a length > 255...

File: rasa/engine/storage/local_model_storage.py [hunk 1]
```diff
@@ -26,6 +27,32 @@
 MODEL_ARCHIVE_METADATA_FILE = "metadata.json"
 
 
+@contextmanager
+def windows_safe_temporary_directory(
```
Comment (m-vdb) [inline rasa/engine/storage/local_model_storage.py:31]: from a code organisation standpoint, not sure if this should be in another utils file or stay here 🤔

Comment (tmbo) [reply rasa/engine/storage/local_model_storage.py:31]: since it is only used in here, i don't see the need to move it somewhere else

**Review by tmbo — APPROVED** [review]
lgtm - thanks a lot for digging into this one 🔥

**Comment by tmbo** [comment]
before you merge, please add a changelog entry.

**Comment by tmbo** [comment]
this should be released together with #11819 as that should hopefully be ready tomorrow as well

**Pull request merged**